### PR TITLE
Fix filesystem

### PR DESCRIPTION
**Pull request by joshuafontany**
This pulls all the `filesystem.js` and related updates out of my experimental draft-PR. This should be merged alongside https://github.com/Jermolene/TiddlyWiki5/pull/5416

They key changes are:
* Ensure all callbacks recieve a second argument "fileInfo/adaptorInfo" & do not rely on global objects during async tasks
  * Makes sure that all code-paths return a valid second data-object if the method does not error.
  * Wrap an anon function around the callback when calling certain $tw.utils functions.
* Patch a vulnerability with the final filePath when writing to disk.
  * Hooks should be called first, then it should be tested for the "encode" logic to prevent writing outside the wiki directory.
* In `WikiFolderMaker.prototype.saveTiddler` use the path/ext filters and a try/catch block to test for `saveTiddlerToFileSync` errors and log to the console.
  * This prevents file-save errors from crashing the command.

### CODE REVIEW

File: core/modules/commands/savewikifolder.js [hunk 1]
```diff
@@ -158,11 +158,26 @@ WikiFolderMaker.prototype.saveCustomPlugin = function(pluginTiddler) {
 };
 
 WikiFolderMaker.prototype.saveTiddler = function(directory,tiddler) {
+	var title = tiddler.fields.title, fileInfo, pathFilters, extFilters;
+	if(this.wiki.tiddlerExists("$:/config/FileSystemPaths")){
```
Comment (Jermolene) [inline core/modules/commands/savewikifolder.js:162]: There should be a space between the closing parenthesis and the opening brace

File: core/modules/commands/savewikifolder.js [hunk 2]
```diff
@@ -158,11 +158,26 @@ WikiFolderMaker.prototype.saveCustomPlugin = function(pluginTiddler) {
...
+		pathFilters: pathFilters,
+		extFilters: extFilters,
+		originalpath: this.wiki.extractTiddlerDataItem("$:/config/OriginalTiddlerPaths",title, "")
 	});
-	$tw.utils.saveTiddlerToFileSync(tiddler,fileInfo);
+	try {
+		$tw.utils.saveTiddlerToFileSync(tiddler,fileInfo);
+	} catch (err) {
+		console.log("SaveWikiFolder: Error saving file '" + fileInfo.filepath + "', tiddler: '" + tiddler.fields.title);
+		console.error(err);
```
Comment (Jermolene) [inline core/modules/commands/savewikifolder.js:179]: We should, but we don't currently consistently send error messages to console.error. Perhaps that's something we should do in bulk as a separate PR?

File: core/modules/utils/filesystem.js [hunk 3]
```diff
@@ -419,14 +417,29 @@ exports.saveTiddlerToFile = function(tiddler,fileInfo,callback) {
...
+					return callback(err);
+				}
+				return callback(null,fileInfo);
+			});
 		});
 	} else {
 		// Save the tiddler as a self contained templated file
 		if(fileInfo.type === "application/x-tiddler") {
-			fs.writeFile(fileInfo.filepath,tiddler.getFieldStringBlock({exclude: ["text","bag"]}) + (!!tiddler.fields.text ? "\n\n" + tiddler.fields.text : ""),"utf8",callback);
+			fs.writeFile(fileInfo.filepath,tiddler.getFieldStringBlock({exclude: ["text","bag"]}) + (!!tiddler.fields.text ? "\n\n" + tiddler.fields.text : ""),"utf8",function(err){
```
Comment (Jermolene) [inline core/modules/utils/filesystem.js:430]: Again, there should be another space: `function(err) {`

**Review by Jermolene — CHANGES_REQUESTED** [review]
Thanks @joshuafontany a couple of minor points

File: core/modules/commands/savewikifolder.js (hunk 2 above)
Comment (joshuafontany) [reply core/modules/commands/savewikifolder.js:179]: Agree. I think it would be more consistent to use $tw.logger.alert() here (& in similar cases), and that logger method should print errors to both node and browser console (along with the alert-tiddler logic). That should be a separate PR.

**Comment by joshuafontany** [comment]
@Jermolene Thanks for catching those. :)

**Comment by Jermolene** [comment]
Thanks @joshuafontany
